ledger/participant-state/kvutils/src/main/scala/kvutils/com/daml/ledger/participant-state/kvutils/KeyValueUtils.scala

```diff
@@ -0,0 +1,447 @@
+
+  /** A contract instance along with its ledger effective time from which it is active. */
+  type StampedContractInstance = (Timestamp, ContractInst[Transaction.Value[AbsoluteContractId]])
+
+  /** KVEntry is the serialized 'DamlKVEntry', containing either the submitted transaction and related metadata
+    * or the DAML-LF archive. */
+  type KVEntryBlob = ByteString
+
+  /** The inputs to the transaction. */
+  final case class Inputs(
+      /* The contracts referenced in the transaction.
+       * When committing the activeness of the referenced contracts at the ledger effective time
+       * of the transaction must be validated.
+       */
+      contracts: List[KVAbsoluteContractId],
+      /* FIXME(JM): Contract key inputs */
+
+      /* The DAML-LF packages referenced in the transaction. */
+      // FIXME(JM): We need this information from DAML Engine!
+      packages: List[PackageId]
+  )
+
+  /** The effects of the transaction, that is what contracts
+    * were consumed and created, and what contract keys were updated.
+    */
+  case class Effects(
+      /** The contracts consumed by this transaction.
+        * When committing the transaction these contracts must be marked consumed.
+        * A contract should be marked consumed when the transaction is committed,
+        * regardless of the ledger effective time of the transaction (e.g. a transaction
+        * with an earlier ledger effective time that gets committed later would find the
+        * contract inactive).
+        */
+      consumedContracts: List[KVAbsoluteContractId],
+      /** The contracts created by this transaction.
+        * When the transaction is committed, keys marking the activeness of these
+        * contracts should be created. The key should be a combination of the transaction
+        * id and the relative contract id (that is, the node index).
+        */
+      createdContracts: List[KVRelativeContractId]
+
+      // FIXME(JM): updated contract keys
+  )
+
+  /** Transaction information required for validation, consistency checks
+    * and for updating contract activeness. */
+  case class TxInfo(
+      /** The party that submitted the transaction. */
+      submitter: Party,
+      /** The ledger effective time of the transaction. The activeness of the declared inputs should
+        * be checked in relation to this (except if archived, see comment near [[Effects.consumedContracts]]).
+        */
+      ledgerEffectiveTime: Timestamp,
+      /** The inputs to the transaction. */
+      inputs: Inputs,
+      /** The effects of the transaction: the contracts consumed and created. */
+      effects: Effects,
+      /** The decoded transaction. */
+      tx: SubmittedTransaction,
+  )
+
+  /** Given a KVTransaction and a contract id, deserialize the transaction and
+    * and produce the absolute contract instance.
+    *
+    * FIXME(JM): We should cache some of the work to deserialize the transactions.
```

Comment:
Why? It is cached by the `Indexer`; and we can't really do that in an actual transaction-processor implementation can we?

Reply:
this is used by validation, so index service isn't available.